engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala

```diff
@@ -184,20 +215,79 @@ class WorkflowActor(workflowId: WorkflowId,
   when(WorkflowFailedState) { FSM.NullFunction }
   when(WorkflowSucceededState) { FSM.NullFunction }
 
-  onTransition {
-    case oldState -> terminalState if terminalState.terminal =>
-      log.info(s"$tag transition from $oldState to $terminalState: shutting down")
-      context.stop(self)
-    case fromState -> toState =>
-      log.info(s"$tag transitioning from $fromState to $toState")
-  }
-
   whenUnhandled {
+    case Event(MetadataServiceFailureResponse(error), _) =>
+      // Do something useful here??
+      log.warning(s"$tag Received MetadataServiceFailure")
+      stay
+    case Event(MetadataActionFailureResponse(error, action), _) =>
+      // Do something useful here??
+      log.warning(s"$tag Received MetadataServiceFailure")
+      stay
     case Event(AbortWorkflowCommand, WorkflowActorData(Some(actor), _)) =>
       actor ! EngineLifecycleActorAbortCommand
       goto(AbortingWorkflowState)
     case unhandledMessage =>
       log.warning(s"$tag received an unhandled message $unhandledMessage in state $stateName")
       stay
   }
+
+  onTransition {
+    case fromState -> toState =>
+      log.info(s"$tag transitioning from $fromState to $toState")
+      // This updates the workflow status
+      pushCurrentStateToMetadataService(toState.workflowState) onFailure {
+        case reason: Throwable => self ! MetadataServiceFailureResponse(reason)
+      }
+  }
+
+  onTransition {
+    case oldState -> terminalState if terminalState.terminal =>
+      log.info(s"$tag transition from $oldState to $terminalState: shutting down")
+      // Add the end time of the workflow in the MetadataService
+      val metadataEventMsg = MetadataEvent(MetadataKey(workflowId, None, WorkflowMetadataKeys.EndTime), MetadataValue(DateTime.now.toString))
+      serviceRegistryActor.publishToMetadataService(metadataEventMsg) map handleMetadataServiceResponse recover {
```

Comment:
How this recover is different from recover in publishToMetadataService()... ? I think you were using that recover only for logging which can be done here as well.